bootstore/src/schemes/v0/peer_networking.rs

```diff
@@ -0,0 +1,635 @@
+                    self.close().await
+                } else {
+                    match write_framed(&msg) {
+                        Ok(msg) => {
+                            self.write_queue.push_back(msg);
+                            Ok(())
+                        }
+                        Err(e) => {
+                            warn!(
+                                self.log,
+                                "Closing connection: Failed to serialize msg: {}",
+                                e
+                            );
+                            self.close().await
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    async fn check_write_result(
+        &mut self,
+        res: Result<usize, std::io::Error>,
+    ) -> Result<(), ConnErr> {
+        match res {
+            Ok(_) => {
+                if !self.current_write.has_remaining() {
+                    self.current_write = Cursor::new(Vec::new());
+                }
+                Ok(())
+            }
+            Err(e) => {
+                warn!(self.log, "Closing connection: Failed to write: {}", e);
+                self.close().await
+            }
+        }
+    }
+
+    async fn read(
+        &mut self,
+        res: Result<usize, std::io::Error>,
+    ) -> Result<(), ConnErr> {
+        match res {
+            Ok(n) => {
+                self.total_read += n;
+            }
+            Err(e) => {
+                warn!(self.log, "Closing connection: failed to read: {}", e);
+                return self.close().await;
+            }
+        }
+
+        // We may have more than one message that has been read
+        loop {
+            if self.total_read < FRAME_HEADER_SIZE {
+                return Ok(());
+            }
+            // Read frame size
+            let size = read_frame_size(
+                self.read_buf[..FRAME_HEADER_SIZE].try_into().unwrap(),
+            );
+            let end = size + FRAME_HEADER_SIZE;
+            if end > self.total_read {
+                return Ok(());
+            }
+            let msg: Msg = match ciborium::from_reader(
+                &self.read_buf[FRAME_HEADER_SIZE..end],
+            ) {
+                Ok(msg) => {
+                    // Move any remaining bytes to the beginning of the buffer.
+                    self.read_buf.copy_within(end..self.total_read, 0);
+                    self.total_read = self.total_read - end;
+                    msg
+                }
+                Err(e) => {
+                    warn!(
+                        self.log,
+                        "Closing connection: failed to deserialize: {}", e
+                    );
+                    return self.close().await;
+                }
+            };
+            self.last_received_msg = Instant::now();
+            debug!(self.log, "Received {:?}", msg);
+            if let Msg::Fsm(msg) = msg {
```

Comment:
This begs the question of what other things `mgs` could be, and why we don't need to handle them.  I'd suggest an explicit `match`, e.g.
```rust
match msg {
    Msg::Fsm(mgs) => {
        // [same as before]
    }
    Msg::Ping => {
        // Nothing to do here, since Ping is just to keep us alive and
        // we updated self.last_received_msg above.
    }
}
```

Reply:
Done in 851cc853f